sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java

```diff
@@ -151,194 +156,216 @@ protected String getLoggerTag() {
                     @Override
                     public void callCosmosDb(final TokenResult tokenResult) {
-                        // https://docs.microsoft.com/en-us/rest/api/cosmos-db/get-a-document
-                        CosmosDb.callCosmosDb(
-                                tokenResult,
-                                documentId,
-                                createHttpClient(mContext),
-                                METHOD_GET,
-                                "",
-                                new ServiceCallback() {
-
-                                    @Override
-                                    public void onCallSucceeded(final String payload, Map<String, String> headers) {
-                                        result.complete(new Document<T>(payload, tokenResult.partition(), documentId));
-                                    }
-
-                                    @Override
-                                    public void onCallFailed(Exception e) {
-                                        handleApiCallFailure(e);
-                                        result.complete(new Document<T>(e));
-                                    }
-                                });
-
-                        // TODO: do we need to call `complete` here?
+                        callCosmosDbReadApi(tokenResult, documentId, result);
                     }
 
                     @Override
                     public void completeFuture(Exception e) {
-                        result.complete(new Document<T>(e));
-                    }
-                });
-
+                        completeFutureAndRemovePendingCall(e, result);
+                    }});
         return result;
     }
 
-    private TokenResult parseTokenResult(String payload) {
-        TokensResponse tokensResponse = gson.fromJson(payload, TokensResponse.class);
-        return tokensResponse.tokens().get(0);
+    private synchronized <T> void callCosmosDbReadApi(final TokenResult tokenResult, final String documentId, final DefaultAppCenterFuture<Document<T>> result) {
+        ServiceCall cosmosDbCall = CosmosDb.callCosmosDbApi(
+                tokenResult,
+                documentId,
+                mHttpClient,
+                METHOD_GET,
+                null,
+                new ServiceCallback() {
+                    @Override
+                    public void onCallSucceeded(String payload, Map<String, String> headers) {
+                        completeFutureAndRemovePendingCall(new Document<T>(payload, tokenResult.partition(), documentId), result);
+                    }
+
+                    @Override
+                    public void onCallFailed(Exception e) {
+                        completeFutureAndRemovePendingCall(e, result);
+                    }});
+        mPendingCalls.put(result, cosmosDbCall);
     }
 
+    //endregion
+
+    //region List implementation
+
     // List (need optional signature to configure page size)
```

Comment:
all comments I meant for `/*`, I commented only in 1 place but was global feedback.

Also do region comments have to be `//` to work?

Reply:
Yeah, it's an IDE thing. I don't think it would work with /* comments. It's not a comment, technically, but an instruction to IDE to be able to collapse and expand sections of code.